pkg/scheduling/v2/scheduler.go

```diff
@@ -0,0 +1,602 @@
+
+		if !rlResult.succeeded {
+			res.rateLimitResult = &rlResult
+			return res, nil
+		}
+	}
+
+	// pick a worker to assign the slot to
+	var assignedSlot *slot
+
+	s.actionsMu.RLock()
+	if _, ok := s.actions[actionId]; !ok {
+		s.actionsMu.RUnlock()
+		res.noSlots = true
+		return res, nil
+	}
+
+	candidateSlots := s.actions[actionId].slots
+	s.actionsMu.RUnlock()
+
+	candidateSlots = getRankedSlots(qi, labels, candidateSlots)
+
+	for i := skip; i < len(candidateSlots); i++ {
+		slot := candidateSlots[i]
+
+		if !slot.active() {
+			continue
+		}
+
+		if !slot.use([]func(){rateLimitAck}, []func(){rateLimitNack}) {
+			continue
+		}
+
+		assignedSlot = slot
+		break
+	}
+
+	if assignedSlot == nil {
+		res.noSlots = true
+		return res, nil
+	}
+
+	s.assignedCountMu.Lock()
+	s.assignedCount++
+	res.ackId = s.assignedCount
+	s.assignedCountMu.Unlock()
+
+	s.unackedMu.Lock()
+	s.unackedSlots[res.ackId] = assignedSlot
+	s.unackedMu.Unlock()
+
+	res.workerId = sqlchelpers.UUIDFromStr(assignedSlot.getWorkerId())
+	res.succeeded = true
+
+	return res, nil
+}
+
+type AssignedQueueItem struct {
+	AckId    int
+	WorkerId pgtype.UUID
+
+	QueueItem *dbsqlc.QueueItem
+
+	// DispatcherId only gets set after a successful flush to the database
+	DispatcherId *pgtype.UUID
+}
+
+type erroredQueueItem struct {
+	QueueItem *dbsqlc.QueueItem
+	Err       error
+}
+
+type assignResults struct {
+	assigned           []*AssignedQueueItem
+	errored            []*erroredQueueItem
+	unassigned         []*dbsqlc.QueueItem
+	schedulingTimedOut []*dbsqlc.QueueItem
+	rateLimited        []*rateLimitResult
+}
+
+func (s *Scheduler) tryAssign(
+	ctx context.Context,
+	qis []*dbsqlc.QueueItem,
+	stepIdsToLabels map[string][]*dbsqlc.GetDesiredLabelsRow,
+	stepRunIdsToRateLimits map[string]map[string]int32,
+) <-chan *assignResults {
+	// split into groups based on action ids, and process each action id in parallel
+	actionIdToQueueItems := make(map[string][]*dbsqlc.QueueItem)
+
+	for i := range qis {
+		qi := qis[i]
+
+		actionId := qi.ActionId.String
+
+		if _, ok := actionIdToQueueItems[actionId]; !ok {
+			actionIdToQueueItems[actionId] = make([]*dbsqlc.QueueItem, 0)
+		}
+
+		actionIdToQueueItems[actionId] = append(actionIdToQueueItems[actionId], qi)
+	}
+
+	resultsCh := make(chan *assignResults, len(actionIdToQueueItems))
+
+	go func() {
+		wg := sync.WaitGroup{}
+
+		// process each action id in parallel
+		for actionId, qis := range actionIdToQueueItems {
+			wg.Add(1)
+
+			go func(actionId string, qis []*dbsqlc.QueueItem) {
+				defer wg.Done()
+				assigned := make([]*AssignedQueueItem, 0, len(qis))
+				errored := make([]*erroredQueueItem, 0, len(qis))
+				unassigned := make([]*dbsqlc.QueueItem, 0, len(qis))
+				schedulingTimedOut := make([]*dbsqlc.QueueItem, 0, len(qis))
+				rateLimited := make([]*rateLimitResult, 0, len(qis))
+
+				startAssignment := time.Now()
+
+				// TODO: ADD A NOTE ABOUT SKIPPING SINCE THERE MIGHT BE CONTENTION ON THE CANDIDATE SLOTS
```

Comment:
would be a nice note

Reply:
realized that the skip may be incorrect in certain cases, so turned this into a `ringOffset` which is basically a hint for where to start the iteration.